### PR TITLE
Created test for factory decorator and added scalar type hints

### PR DESCRIPTION
**Pull request by jongotlin**
Issue #17

### CODE REVIEW

File: tests/AbstractFactoryDecoratorTest.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+class AbstractFactoryDecoratorTest extends TestCase
+{
+    public function testConstruct()
+    {
+        $factory = new class extends AbstractFactoryDecorator {
+            public function createNewInstance(?string $raw): IdInterface {}
+        };
+
+        $this->assertInstanceOf(IdFactory::class, $factory);
```
Comment (hanneskod) [inline tests/AbstractFactoryDecoratorTest.php]: This should be `IdFactoryInterface`. The fact that `AbstractFactoryDecorator` subclass `IdFactory` is a blast from the past That should be `implements IdFactoryInterface` in 2.0.

File: tests/AbstractFactoryDecoratorTest.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    public function testCreateId()
+    {
+        $factory = new class extends AbstractFactoryDecorator {
+            public function createNewInstance(?string $raw): IdInterface {
+                return new class extends AbstractId {};
+            }
+        };
+
+        $this->assertInstanceOf(AbstractId::class, $factory->createId('123'));
```
Comment (hanneskod) [inline tests/AbstractFactoryDecoratorTest.php]: Same here. Let's go `IdInterface`. Or better still mock `IdInterface` and `assertSame` here..

File: tests/AbstractFactoryDecoratorTest.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        $factory = new class extends AbstractFactoryDecorator {
+            public function createNewInstance(?string $raw): IdInterface {}
+        };
+
+        $this->assertInstanceOf(IdFactory::class, $factory);
+    }
+
+    public function testCreateId()
+    {
+        $factory = new class extends AbstractFactoryDecorator {
```
Comment (hanneskod) [inline tests/AbstractFactoryDecoratorTest.php]: One could use `getMockForAbstractClass` and return a mock of `IdInterface`. See comment below..

File: tests/AbstractFactoryDecoratorTest.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                return new class extends AbstractId {};
+            }
+        };
+
+        $this->assertInstanceOf(AbstractId::class, $factory->createId('123'));
+    }
+
+    public function testUseNextFactoryIfIdCannotBeCreated()
+    {
+        $nextFactory = $this->createMock(AbstractFactoryDecorator::class);
```
Comment (hanneskod) [inline tests/AbstractFactoryDecoratorTest.php]: IdFactoryInterface

File: tests/AbstractFactoryDecoratorTest.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        $nextFactory->expects($this->once())->method('createId');
```
Comment (hanneskod) [inline tests/AbstractFactoryDecoratorTest.php]: Return something so that we can assert later?

File: src/AbstractFactoryDecorator.php [hunk 6]
```diff
@@ -17,16 +17,16 @@
...
 
     /**
      * Create ID object from raw id string
      *
-     * @param  string $raw Raw id string
+     * @param  string|null $raw Raw id string
      * @return IdInterface
      */
-    public function createId($raw)
+    public function createId(?string $raw): IdInterface
```
Comment (hanneskod) [inline src/AbstractFactoryDecorator.php]: Why do we need this to be nullable? I'm not sure I follow..

File: tests/AbstractFactoryDecoratorTest.php (hunk 3 above)
Comment (jongotlin) [reply tests/AbstractFactoryDecoratorTest.php]: In that case we would assert only on mocks.

File: tests/AbstractFactoryDecoratorTest.php (hunk 5 above)
Comment (jongotlin) [reply tests/AbstractFactoryDecoratorTest.php]: We test that the next factory is called. What it returns is not important.

**Comment by jongotlin** [comment]
Thanks for your review! Updated with correct mocked interfaces added some comments.

**Comment by hanneskod** [comment]
Hi and again sorry for the hold up.. I see your point in not writing tests with only mocks. The way phpunits mocking model works I think the relevant lines in AbstractFactoryDecorator would still be tested, but it would in a way be an odd test anyway, I agree.

I don't think the createId argument should be nullable though. The interface consumes a string so that is how we should declare the type in my opinion. That also makes it possible to use statical analysis to find places where null is being passed. TypeError is good as passing null is an error much in that same way as calling a non-existing method or simliar.. I see no problem there.

**Comment by jongotlin** [comment]
Updated to allow null. Should we leave the test as is or do you want me to change something there?

**Comment by hanneskod** [comment]
Yupp, looks good to me. Thank you! Merging..

**Pull request merged**